examples/factory/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+    /// factory.
+    #[derive(Debug, Serialize, SchemaType, Eq, PartialEq, PartialOrd, Clone)]
+    pub struct Product {
+        /// The factory that created the product.
+        pub factory: ContractAddress,
+        /// The index given to the product by the factory.
+        pub index:   u64,
+    }
+
+    /// The overall state of a product, which may or may not yet have been
+    /// initialized by a factory.
+    #[derive(Debug, Serialize, SchemaType, Eq, PartialEq, PartialOrd, Clone)]
+    pub enum ProductState {
+        /// The product has not yet been initialized by the factory.
+        Uninitialized,
+        /// The product has been initialised by the factory.
+        Initialized(Product),
+    }
+
+    /// Errors
+    #[derive(Serialize, Debug, PartialEq, Eq, Reject, SchemaType)]
+    pub enum ProductError {
+        /// Failed parsing the parameter.
+        #[from(ParseError)]
+        ParseParams,
+        /// The originator of the transaction does not match the owner of the
+        /// instance, and thus is not authorized to initialize the contract.
+        NotAuthorized,
+        /// The product was already initialized.
+        AlreadyInitialized,
+        /// [`initialize`] was not called by a factory contract.
+        NotInitializedByFactory,
+    }
+
+    /// Construct the product contract in the uninitialized state.
+    #[init(contract = "product")]
+    pub fn init(_ctx: &InitContext, _state_builder: &mut StateBuilder) -> InitResult<ProductState> {
+        Ok(ProductState::Uninitialized)
+    }
+
+    /// Initialize the product contract, supplying its index.
+    /// The following preconditions are checked:
+    ///
+    /// * The instance must not have already been initialized. Otherwise,
+    ///   [`ProductError::AlreadyInitialized`] is returned.
+    ///
+    /// * The account invoking the current transaction must be the same account
+    ///   that created the instance. Otherwise [`ProductError::NotAuthorized`]
+    ///   is returned.
+    ///
+    /// * The method must be called by another smart contract (the factory). If
+    ///   the caller is not a contract,
+    ///   [`ProductError::NotInitializedByFactory`] is returned.
+    ///
+    /// Note that beyond checking that the caller is a smart contract, this
+    /// method does not check further properties of the caller (e.g. that it is
+    /// any particular factory smart contract).
+    #[receive(contract = "product", name = "initialize", parameter = "u64", mutable)]
+    pub fn initialize(
+        ctx: &ReceiveContext,
+        host: &mut Host<ProductState>,
+    ) -> Result<(), ProductError> {
+        let state = host.state_mut();
+        // Check that the contract has not already been initialized.
+        if let ProductState::Initialized(_) = *state {
+            Err(ProductError::AlreadyInitialized)?;
+        }
+        // Check that the invoker is the owner of the contract.
+        if ctx.invoker() != ctx.owner() {
```

Comment:
We try to avoid using `ctx.invoker` (`tx_origin` equivalent on Ethereum) for authorization in smart contracts.

https://consensys.github.io/smart-contract-best-practices/development-recommendations/solidity-specific/tx-origin/

`ctx.sender` should be used for authorization.

Reply:
I suggest we don't have this check. If we want the `owner` of the product to initialize the `product` with some custom values then this should be done in the `constructor` logic of the `product` contract.

Is there a specific use case where we have to restrict the `product` to be added to the factory via the `produce` entrypoint by the deployer address of the `product`?

Reply:
The reason for this check is that we don't want a malicious third-party to add the product to a factory (which might, for instance, be a completely different factory than the initiator intended). Ideally, we would like the product to be constructed in the same transaction as the factory produces it. But since that is impossible, ensuring the transactions are from the same origin gets close enough, which is what this achieves.

The alternatives seem to be:

1. We assume that no third party will be sufficiently motivated to DoS attack the process. This could be fine, depending on the application.
2. The constructor logic of the product takes all of the custom initialization. This would include indicating which factory is authorised to produce the product. A third party would then be constrained to only produce the product exactly as the originator intended.
3. The product authorises using `ctx.sender`, but is then responsible for calling `produce` on the factory itself. i.e. instead of `user -> factory.produce -> product.initialize` the flow is `user -> product.call_produce -> factory.produce -> product.initialize`. I think this is a bit too complex.

I think the approach of checking the invoker is relatively clean and simple. It allows all of the initial parameters to be set in the call to `produce`, rather than splitting initialization across the constructor and `initialize`. I agree that in general it could be a mistake to use `ctx.invoker` for authorization, but I think this is a special case.

Reply:
I suggest leaving a comment above the `ctx.invoker() != ctx.owner()` pointing out that in some edge cases `ctx.invoker` shouldn't be used in authorization. An even include your alternative suggestion in a comment in the contract :).

Reply:
Of course this could happen, but I think that is not really worth guarding against. It requires the user to sign off on a scam transaction right between initialising the product and calling `produce`. Practically, the window between those two operations should be kept as small as possible. We also cannot completely prevent a user for falling for scams. But I think generally the result of a product being initialized incorrectly should generally be that the product is not really usable for its intended purpose, rather than it is useful for some malicious purpose. (The ways I could see it being used for some malicious purpose (as a result of this kind of hijacking) is if the ownership of the contract is somehow used for authorization, or if some privileges are granted to the contract before it is initialized.)

I think it's also worth noting that the flow `user -> contract -> factory.produce -> product.initialize` might also be useful in practice, so is not necessarily something we want to prohibit on the basis that it could be used maliciously.

Reply:
I have added an extended discussion in the top-level documentation of this file, which will hopefully address these issues.